File: guess.ts
type int = number;

type Game = {
  answer: int;
  done: boolean;
  guesses: int;
  high: int;
};

let errCount = 0;

function askGuess(high: int): int {
  const text = prompt(`Guess a number between 1 and ${high}:`);
  return parseIntChecked(text);
}

function askGuessMulti(high: int): int {
  while (true) {
    try {
      return askGuess(high);
    } catch {
      console.log("I didn't understand");
      errCount += 1;
    }
  }
}

function parseIntChecked(text: string | null | undefined): int {
  const value = parseInt(text as string);
  if (value != Number(text)) throw new Error(`bad int: ${text}`);
  return value;
}

function pickAnswer(high: int): int {
  return Math.floor(Math.random() * high) + 1;
}

function play(game: Game) {
  while (!game.done) {
    const guess = askGuessMulti(game.high);
    report(game, guess);
    update(game, guess);
  }
}

function report(game: Game, guess: int) {
  // deno-fmt-ignore
  const description =
    guess < game.answer ? "too low" :
    guess > game.answer ? "too high" :
    "the answer!";
  console.log(`${guess} is ${description}`);
}

function update(game: Game, guess: int) {
  if (guess == game.answer) {
    game.done = true;
  }
  game.guesses += 1;
}

export function main() {
  const high = 100;
  const answer = pickAnswer(high);
  const game = { answer, done: false, guesses: 0, high };
  play(game);
  console.log(`Finished in ${game.guesses} guesses`);
  console.debug(`Total input errors: ${errCount}`)
}

main();

File: guess.m
function guess
    global errCount;
    errCount = 0;
    % Set up game.
    game.done = false;
    game.guesses = 0;
    game.high = 100;
    game.answer = randi(game.high);
    % Play and report.
    game = play(game);
    fprintf('Finished in %d guesses\n', game.guesses);
    fprintf('Total input errors: %d\n', errCount);
end

function guess = askGuess(high)
    prompt = sprintf('Guess a number between 1 and %d: ', high);
    guess = parseIntChecked(input(prompt, 's'));
end

function guess = askGuessMulti(high)
    while true
        try
            guess = askGuess(high);
            return;
        catch
            fprintf('I didn''t understand\n');
            global errCount;
            errCount += 1;
        end
    end
end

function value = parseIntChecked(text)
    [value, _, err] = sscanf(text, '%d');
    if ~isempty(err)
        error(sprintf('bad int: %s', text));
    end
end

function game = play(game)
    while ~game.done
        guess = askGuessMulti(game.high);
        report(game, guess);
        game = update(game, guess);
    end
end

function report(game, guess)
    if guess < game.answer
        description = 'too low';
    elseif guess > game.answer
        description = 'too high';
    else
        description = 'the answer!';
    end
    fprintf('%d is %s\n', guess, description);
end

function game = update(game, guess)
    if guess == game.answer
        game.done = true;
    end
    game.guesses += 1;
end
